service/service.go

```diff
@@ -1,6 +1,64 @@
 package service
 
-import "release-manager/service/model"
+import (
+	"context"
+
+	cryptox "release-manager/pkg/crypto"
+	"release-manager/service/model"
+
+	"github.com/google/uuid"
+)
+
+type authRepository interface {
+	ReadUserIDForToken(ctx context.Context, token string) (uuid.UUID, error)
+}
+
+type environmentRepository interface {
+	Create(ctx context.Context, env model.Environment) error
+	Read(ctx context.Context, envID uuid.UUID) (model.Environment, error)
+	ReadByNameForProject(ctx context.Context, projectID uuid.UUID, name string) (model.Environment, error)
+	ReadAllForProject(ctx context.Context, projectID uuid.UUID) ([]model.Environment, error)
+	Delete(ctx context.Context, envID uuid.UUID) error
+	Update(ctx context.Context, env model.Environment) error
+}
+
+type projectRepository interface {
+	Create(ctx context.Context, p model.Project) error
+	Read(ctx context.Context, id uuid.UUID) (model.Project, error)
+	ReadAll(ctx context.Context) ([]model.Project, error)
+	Update(ctx context.Context, p model.Project) error
+	Delete(ctx context.Context, id uuid.UUID) error
+}
+
+type userRepository interface {
+	Read(ctx context.Context, id uuid.UUID) (model.User, error)
+	ReadAll(ctx context.Context) ([]model.User, error)
+	Delete(ctx context.Context, id uuid.UUID) error
+}
+
+type settingsRepository interface {
+	Update(ctx context.Context, c model.Settings) error
+	Read(ctx context.Context) (model.Settings, error)
+}
+
+type projectInvitationRepository interface {
+	Create(ctx context.Context, i model.ProjectInvitation) error
+	Read(ctx context.Context, id uuid.UUID) (model.ProjectInvitation, error)
+	ReadByEmailForProject(ctx context.Context, email string, projectID uuid.UUID) (model.ProjectInvitation, error)
+	ReadByTokenHashAndStatus(ctx context.Context, hash cryptox.Hash, status model.ProjectInvitationStatus) (model.ProjectInvitation, error)
+	ReadAllForProject(ctx context.Context, projectID uuid.UUID) ([]model.ProjectInvitation, error)
+	Update(ctx context.Context, i model.ProjectInvitation) error
+	Delete(ctx context.Context, id uuid.UUID) error
+}
+
+type authGuard interface {
+	AuthorizeAdminRole(ctx context.Context, userID uuid.UUID) error
+	AuthorizeRole(ctx context.Context, userID uuid.UUID, role model.UserRole) error
+}
+
+type projectGetter interface {
+	Get(ctx context.Context, projectID, authUserID uuid.UUID) (model.Project, error)
+}
```

Comment:
This is just temporary because the `Project` and `ProjectInvitation` services will be merged into one, and this interface will no longer be necessary.

I borrowed name `projectGetter` from [this book](https://www.amazon.com/100-Mistakes-How-Avoid-Them/dp/1617299596).

I initially felt it was overly descriptive. However, It improved code readability: (and we were actually discussing exactly this call)

```
if _, err := s.projectGetter.Get(....); err != nil {

}
```

Reply:
>This is just temporary because the Project and ProjectInvitation services will be merged into one, and this interface will no longer be necessary.

💯 

>I borrowed name projectGetter from [this book](https://www.amazon.com/100-Mistakes-How-Avoid-Them/dp/1617299596).

Yeah, a must-read book that I.. haven't read yet! 😁 
The more interesting it will be during reviews - as always, feel free to counter-argue me. 😈 

This one is one of my favorite topics, actually. 🔥 
Academically, I think you are absolutely correct.
1) You are using interfaces
2) You are keeping them as slim and explicit as possible
So, the resulting potential for decoupling etc. is maximum or almost maximum.

I once tried to promote a similar thing at my project as well, but the counter-argument was:
- Yes, but then it's too scattered, too complex. Unnecessarily scattered and complex, because we know FOR SURE that it's going to be (in this case) `ProjectService` who will implement `Get()` method, `Delete()` method etc. In other words, we are practically guaranteed to not have a `ProjectGetterService` and `ProjectDeleterService`.

If we take this argument to extreme, we might say - well, if we are guaranteed to have one and only one implementation, one `ProjectService`, why do we need an interface here at all?
If you are really into that, I won't even spoil the "answers". 😁 
But one of them can be nicely and clearly inferred from the classic [Uncle Bob's talk about SOLID principles](https://www.youtube.com/watch?v=zHiWqnTWsn4). Must watch.

Reply:
> Yeah, a must-read book that I.. haven't read yet! 😁

I havent read the whole book actually yet. Some chapters I have found too heavy for me yet as I am new to Go :D. However, I did enjoy the chapters focused on Code and Project Organization.

And I actually agree with you. I think the fact I need interface `projectGetter` more indicates that it would be better to structure service differently. However, in some cases, having a slim interface makes sense to me, such as with `projectInvitationSender` in #30.